### PR TITLE
[CBRD-22980] Replicate serial updates as sub-transactions

### PR DESCRIPTION
**Pull request by ribram**
http://jira.cubrid.org/browse/CBRD-22980

Rewrite serial_update_serial_object and add sub-transactions apply on slave.

#### serial_update_serial_object

- add a new class *serial_heap_record*, to manage reads/updates of serial heap record. it merges common operations from serial_update_cur_val_of_serial and xserial_get_next_value_internal.
- serial_heap_record manages replication and logging too. to properly manage replication, it also handles setting values in heap_cache_attrinfo.
- `serial_heap_record::load` reads record from heap and initializes heap_cache_attrinfo.
- `serial_heap_record::update_current_value` and `serial_heap_record::set_started_value` set values in heap_cache_attrinfo (using update_value private function). replication is automatically started.
- `serial_heap_record::update_record` updates the record in heap; it also "commits" replication.
- serial_heap_record destructor makes sure no resources are leaked.
- replication - there two modes of replication: instant and normal. instant uses `cubreplication::subtran_generate`, normal uses transaction descriptor. replication is managed by start_replication, commit_replication and abort_replication functions. on SA_MODE it is disabled.
- rewrite serial_update_cur_val_of_serial and xserial_get_next_value_internal using serial_heap_record
- serial_update_serial_object no longer manages replication, only heap update and transaction logging

### CODE REVIEW

**Comment by ribram** [comment]
If state/bitmask changes are accepted, http://jira.cubrid.org/browse/CBRD-22937 becomes obsolete.

File: src/replication/replication_subtran_apply.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  {
+    task *new_task = alloc_task (se);
+
+    std::unique_lock<std::mutex> ulock (m_tasks_mutex);
+    if (m_tasks.empty ())
+      {
+	// no preceding tasks, can be pushed
+	m_lc.push_task (new_task);
+      }
+    m_tasks.push_back (new_task);
```
Comment (ribram) [inline src/replication/replication_subtran_apply.cpp]: Question: is this the right place to push the task?

File: src/query/serial.c [hunk 2]
```diff
@@ -1439,3 +1330,295 @@ serial_get_index_btid (BTID * output)
...
+            {
+              heap_attrinfo_end (m_thread_p, &m_attrinfo);
+            }
+        }
+    }
+  else
+    {
+      if (er_errid () == ER_PB_BAD_PAGEID)
+	{
+          error_code = ER_HEAP_UNKNOWN_OBJECT;
```
Comment (eseokoh) [inline src/query/serial.c]: please confirm indentation.

File: src/query/serial.c [hunk 3]
```diff
@@ -1439,3 +1330,304 @@ serial_get_index_btid (BTID * output)
...
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, error_code, 3, OID_AS_ARGS (&m_serial_oid));
+	}
+      else
+	{
+          error_code = ER_QPROC_CANNOT_FETCH_SERIAL;
```
Comment (eseokoh) [inline src/query/serial.c]: please also confirm indentation.

**Pull request closed without merging**